src/com/linkedin/parseq/BaseTask.java

```diff
@@ -389,18 +391,24 @@ private void appendTaskStackTrace(final Throwable error) {
     error.setStackTrace(concatenatedStackTrace);
   }
 
-  protected boolean transitionRun(final TraceBuilder traceBuilder) {
+  protected boolean transitionRun(final Context context) {
     State state;
     State newState;
     do {
       state = _stateRef.get();
       if (state.getType() != StateType.INIT) {
+        // prevent cross-plan task sharing if enabled
+        if (!ParSeqGlobalConfiguration.isAllowCrossPlanSharingEnabled() &&
+            _startPlanId != null && !_startPlanId.equals(context.getPlanId())) {
+          throw new IllegalStateException(this.toString() + " should not be shared across plans!!");
```

Comment:
We can't just throw exception here because it might leave Task in inconsistent state. We also need to log error when this problem occurs. It would be best to change exception to be a checked exception and wrap all calls to this method with try-catch to make sure Task's state is recovered.

Reply:
Added a checked exception class CrossPlanTaskSharingException.